include/coap2/coap_session_internal.h

```diff
@@ -24,19 +28,267 @@
+ *
+ * @return A new CoAP session or NULL if failed. Call coap_session_release to
+ * add to unused queue.
+ */
+coap_session_t *coap_new_server_session(
+  struct coap_context_t *ctx,
+  coap_endpoint_t *ep
+);
+
+/**
+ * Function interface for datagram data transmission. This function returns
+ * the number of bytes that have been transmitted, or a value less than zero
+ * on error.
+ *
+ * @param session          Session to send data on.
+ * @param data             The data to send.
+ * @param datalen          The actual length of @p data.
+ *
+ * @return                 The number of bytes written on success, or a value
+ *                         less than zero on error.
+ */
+ssize_t coap_session_send(coap_session_t *session,
+  const uint8_t *data, size_t datalen);
+
+/**
+ * Function interface for stream data transmission. This function returns
+ * the number of bytes that have been transmitted, or a value less than zero
+ * on error. The number of bytes written may be less than datalen because of
+ * congestion control.
+ *
+ * @param session          Session to send data on.
+ * @param data             The data to send.
+ * @param datalen          The actual length of @p data.
+ *
+ * @return                 The number of bytes written on success, or a value
+ *                         less than zero on error.
+ */
+ssize_t coap_session_write(coap_session_t *session,
+  const uint8_t *data, size_t datalen);
+
+/**
+ * Send a pdu according to the session's protocol. This function returns
+ * the number of bytes that have been transmitted, or a value less than zero
+ * on error.
+ *
+ * @param session          Session to send pdu on.
+ * @param pdu              The pdu to send.
+ *
+ * @return                 The number of bytes written on success, or a value
+ *                         less than zero on error.
+ */
+ssize_t coap_session_send_pdu(coap_session_t *session, coap_pdu_t *pdu);
+
+ssize_t
+coap_session_delay_pdu(coap_session_t *session, coap_pdu_t *pdu,
+                       struct coap_queue_t *node);
+
+/**
+ * Lookup the server session for the packet received on an endpoint, or create
+ * a new one.
+ *
+ * @param endpoint Active endpoint the packet was received on.
+ * @param packet Received packet.
+ * @param now The current time in ticks.
+ * @return The CoAP session or @c NULL if error.
+ */
+coap_session_t *coap_endpoint_get_session(coap_endpoint_t *endpoint,
+  const struct coap_packet_t *packet, coap_tick_t now);
+
+/**
+ * Create a new DTLS session for the @p session.
+ * Note: the @p session is released if no DTLS server session can be created.
+ *
+ * @ingroup dtls_internal
+ *
+ * @param session   Session to add DTLS session to
+ * @param now       The current time in ticks.
+ *
+ * @return CoAP session or @c NULL if error.
+ */
+coap_session_t *coap_session_new_dtls_session(coap_session_t *session,
+  coap_tick_t now);
+
+coap_session_t *coap_session_get_by_peer(struct coap_context_t *ctx,
```

Comment:
Why is this function made internal? It is very helpful in applications.

Reply:
Fair point - will move it back to being visible.